OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs

```diff
@@ -0,0 +1,140 @@
+		{
+			// Get a spawn cell
+			var location = GetNearestInvisibleCell(self.Owner, self.Location);
+			var pos = self.World.Map.CenterOfCell(location);
+
+			var initialFacing = self.World.Map.FacingBetween(location, self.Location, 0);
+
+			// If aircraft, spawn at cruise altitude
+			var ai = producee.Traits.WithInterface<AircraftInfo>().First();
+			if (ai != null)
+				pos += new WVec(0, 0, ai.CruiseAltitude.Range);
+
+			self.World.AddFrameEndTask(w =>
+			{
+				var td = new TypeDictionary
+				{
+					new OwnerInit(self.Owner),
+					new LocationInit(location),
+					new CenterPositionInit(pos),
+					new FacingInit(initialFacing)
+				};
+
+				if (raceVariant != null)
+					td.Add(new RaceInit(raceVariant));
+
+				var newUnit = self.World.CreateActor(producee.Name, td);
+
+				var move = newUnit.TraitOrDefault<IMove>();
+				if (move != null)
+					newUnit.QueueActivity(move.MoveIntoWorld(newUnit, self.Location));
+
+				newUnit.SetTargetLine(Target.FromCell(self.World, self.Location), Color.Green, false);
+
+				if (!self.IsDead)
+					foreach (var t in self.TraitsImplementing<INotifyProduction>())
+						t.UnitProduced(self, newUnit, self.Location);
+
+				var notifyOthers = self.World.ActorsWithTrait<INotifyOtherProduction>();
+				foreach (var notify in notifyOthers)
+					notify.Trait.UnitProducedByOther(notify.Actor, self, newUnit);
+
+				var bi = newUnit.Info.Traits.GetOrDefault<BuildableInfo>();
+				if (bi != null && bi.InitialActivity != null)
+					newUnit.QueueActivity(Game.CreateObject<Activity>(bi.InitialActivity));
+
+				foreach (var t in newUnit.TraitsImplementing<INotifyBuildComplete>())
+					t.BuildingComplete(newUnit);
+			});
+
+			return true;
+		}
+
+		// Gives a (1) cell that is invisible due to fog, or if none (2) a cell that has not been explored yet (shrouded) or as a last resort (3) a cell on map edge.
+		public static CPos GetNearestInvisibleCell(Player player, CPos toCell)
+		{
+			// Get list of players against wich a shroud/fog test should be done
+			var checkplayers = player.World.Players
+				.Where(p => p.Playable == true);
+
+			// Please note hack: World.LobbyInfo.GlobalSettings.Fog returns wrong value for shellmap. Therefore skip to edge spawn on shellmap.
+
+			// Get the closest fogged cell
+			if (player.Shroud.FogEnabled && player.World.Type == WorldType.Regular)
+			{
+				var location = player.World.Map.Cells.Where(c => CheckFogged(c, checkplayers) == true).OrderBy(c => (c - toCell).LengthSquared).FirstOrDefault();
+
+				if (location != CPos.Zero)
+					return location;
+			}
+
+			// Get the closest shrouded cell
+			if (player.Shroud.ShroudEnabled && player.World.Type == WorldType.Regular)
+			{
+				var location = player.World.Map.Cells.Where(c => CheckUnexplored(c, checkplayers) == true).OrderBy(c => (c - toCell).LengthSquared).FirstOrDefault();
+
+				if (location != CPos.Zero)
+					return location;
+			}
+
+			// Last resort
```

Comment:
This is understood as the last attempt to give a valid location (as it is the final return statement).